File: algorithm/113.路径总和.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
  const res = [];
  if (!root) return res;
  function _pathSum(node, num, path) {
    if (!node.left && !node.right) {
      if (!num) {
        res.push(path.split(','));
      }
    } else {
      if (node.left) {
        _pathSum(node.left, num - node.left.val, path + ',' + node.left.val);
      }
      if (node.right) {
        _pathSum(node.right, num - node.right.val, path + ',' + node.right.val);
      }
    }
  }
  _pathSum(root, sum - root.val, root.val.toString());
  return res;
};

File: algorithm/113.路径总和.cpp
#include "iostream"
#include "string"
#include "vector"
using namespace std;

/**
 * 
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
  public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
      vector<vector<int>> res;
      if (!root) return res;
      if (!root -> left && !root -> right) {
        if (sum - (root -> val) == 0) {
          vector<int> temp;
          temp.push_back(root -> val);
          res.push_back(temp);
          return res;
        }
      } else {
        if (root -> left) {
          vector<vector<int>> leftPath = pathSum(root -> left, sum - (root -> val));
          for (int i = 0; i < leftPath.size(); i++) {
            vector<int> temp = leftPath[i];
            vector<int> temp2;
            temp2.push_back(root -> val);
            for (int j = 0; j < temp.size(); j++) {
              temp2.push_back(temp[j]);
            }
            res.push_back(temp2);
          }
        }
        if (root -> right) {
          vector<vector<int>> rightPath = pathSum(root -> right, sum - (root -> val));
          for (int i = 0; i < rightPath.size(); i++) {
            vector<int> temp = rightPath[i];
            vector<int> temp2;
            temp2.push_back(root -> val);
            for (int j = 0; j < temp.size(); j++) {
              temp2.push_back(temp[j]);
            }
            res.push_back(temp2);
          }
        }
      }
      return res;
    }
};
